src/storage/src/render/persist_sink.rs

```diff
@@ -647,126 +656,126 @@ where
+                                    in-flight-batches: {:?}, \
+                                    desired_frontier: {:?}, \
+                                    batch_descriptions_frontier: {:?}",
+                                data,
+                                in_flight_batches,
+                                desired_frontier,
+                                batch_descriptions_frontier,
+                            );
                         }
-                        Event::Progress(frontier) => {
-                            desired_frontier = frontier;
+
+                        let minimum_batch_updates =
+                            persist_clients.cfg().storage_sink_minimum_batch_updates();
+                        for (row, ts, diff) in data {
+                            if write.upper().less_equal(&ts) {
+                                let builder = stashed_batches
+                                    .entry(ts)
+                                    .or_insert_with(BatchBuilderAndMetadata::new);
+
+                                let is_value = row.is_ok();
+
+                                builder
+                                    .add(
+                                        minimum_batch_updates,
+                                        || write.builder(operator_batch_lower.clone()),
+                                        SourceData(row),
+                                        (),
+                                        ts,
+                                        diff,
+                                    )
+                                    .await;
+
+                                source_statistics.inc_updates_staged_by(1);
+
+                                // Note that we assume `diff` is either +1 or -1 here, being anything
+                                // else is a logic bug we can't handle at the metric layer. We also
+                                // assume this addition doesn't overflow.
+                                match (is_value, diff.is_positive()) {
+                                    (true, true) => builder.metrics.inserts += diff.unsigned_abs(),
+                                    (true, false) => {
+                                        builder.metrics.retractions += diff.unsigned_abs()
+                                    }
+                                    (false, true) => {
+                                        builder.metrics.error_inserts += diff.unsigned_abs()
+                                    }
+                                    (false, false) => {
+                                        builder.metrics.error_retractions += diff.unsigned_abs()
+                                    }
+                                }
+                            }
                         }
                     }
-                }
-                else => {
-                    // All inputs are exhausted, so we can shut down.
-                    return;
```

Comment:
we're previously exiting based on the input stream exhaustion, and now based on the frontiers advancing to the empty frontier. Is there a possibility for an issue there? Is it guaranteed that a `Progress` event will be emitted for the empty frontier as the last event of each stream?

Reply:
>  Is it guaranteed that a Progress event will be emitted for the empty frontier as the last event of each stream?

Yes, in fact this is exactly the condition under which the stream returns `None`. https://github.com/MaterializeInc/materialize/blob/v0.108.4/src/timely-util/src/builder_async.rs#L171